File: algorithms/11_golang/graphs/lca_binary_lifting.go
package main

import "fmt"

type LCA struct {
	up    [][]int
	depth []int
	log   int
}

func NewLCA(adj [][]int, root int) *LCA {
	n := len(adj)
	log := 0
	for 1<<log <= n {
		log++
	}
	up := make([][]int, n)
	for i := range up {
		up[i] = make([]int, log)
	}
	depth := make([]int, n)
	var dfs func(u, p int)
	dfs = func(u, p int) {
		up[u][0] = p
		for k := 1; k < log; k++ {
			up[u][k] = up[up[u][k-1]][k-1]
		}
		for _, v := range adj[u] {
			if v != p {
				depth[v] = depth[u] + 1
				dfs(v, u)
			}
		}
	}
	dfs(root, root)
	return &LCA{up: up, depth: depth, log: log}
}

func (l *LCA) Query(a, b int) int {
	if l.depth[a] < l.depth[b] {
		a, b = b, a
	}
	diff := l.depth[a] - l.depth[b]
	for k := 0; k < l.log; k++ {
		if diff&(1<<k) != 0 {
			a = l.up[a][k]
		}
	}
	if a == b {
		return a
	}
	for k := l.log - 1; k >= 0; k-- {
		if l.up[a][k] != l.up[b][k] {
			a = l.up[a][k]
			b = l.up[b][k]
		}
	}
	return l.up[a][0]
}

func main() {
	adj := [][]int{
		{1, 2},
		{0, 3, 4},
		{0, 5, 6},
		{1}, {1}, {2}, {2},
	}
	l := NewLCA(adj, 0)
	if l.Query(3, 4) != 1 || l.Query(3, 6) != 0 || l.Query(5, 6) != 2 {
		panic("LCA: query mismatch")
	}
	fmt.Println("[Go LCA] Binary lifting verified.")
}

File: algorithms/02_c/graphs/lca_binary_lifting.c
#include <stdio.h>

#define MAXN 100000
#define LOG 17

static int up[LOG][MAXN];
static int depth[MAXN];
static int head[MAXN], to[2 * MAXN], next_e[2 * MAXN];
static int edge_cnt;

static void add_edge(int u, int v) {
    to[edge_cnt] = v; next_e[edge_cnt] = head[u]; head[u] = edge_cnt++;
    to[edge_cnt] = u; next_e[edge_cnt] = head[v]; head[v] = edge_cnt++;
}

static void dfs(int u, int parent) {
    up[0][u] = parent;
    for (int k = 1; k < LOG; ++k)
        up[k][u] = (up[k - 1][u] == -1) ? -1 : up[k - 1][up[k - 1][u]];
    for (int e = head[u]; e != -1; e = next_e[e]) {
        int v = to[e];
        if (v == parent) continue;
        depth[v] = depth[u] + 1;
        dfs(v, u);
    }
}

int lca(int u, int v) {
    if (depth[u] < depth[v]) { int t = u; u = v; v = t; }
    int diff = depth[u] - depth[v];
    for (int k = 0; k < LOG; ++k)
        if (diff & (1 << k)) u = up[k][u];
    if (u == v) return u;
    for (int k = LOG - 1; k >= 0; --k)
        if (up[k][u] != up[k][v]) { u = up[k][u]; v = up[k][v]; }
    return up[0][u];
}

int main(void) {
    edge_cnt = 0;
    for (int i = 0; i < MAXN; ++i) head[i] = -1;

    /* tree: 0 root, 0-1-{3,4}, 0-2-{5,6} */
    add_edge(0, 1);
    add_edge(0, 2);
    add_edge(1, 3);
    add_edge(1, 4);
    add_edge(2, 5);
    add_edge(2, 6);

    depth[0] = 0;
    dfs(0, -1);

    if (lca(3, 4) != 1 || lca(3, 5) != 0 || lca(5, 6) != 2) {
        printf("[C LCA] FAILED: binary lifting mismatch\n");
        return 1;
    }
    printf("[C LCA] Binary lifting LCA verified\n");
    return 0;
}
